src/script/lua_api/l_object.cpp

```diff
@@ -1744,26 +1869,250 @@ int ObjectRef::l_get_sky(lua_State *L)
+}
+
+//get_sun(self)
+int ObjectRef::l_get_sun(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	const SunParams &sun_params = player->getSunParams();
+
+	lua_newtable(L);
+	lua_pushboolean(L, sun_params.visible);
+	lua_setfield(L, -2, "visible");
+	lua_pushstring(L, sun_params.texture.c_str());
+	lua_setfield(L, -2, "texture");
+	lua_pushstring(L, sun_params.tonemap.c_str());
+	lua_setfield(L, -2, "tonemap");
+	lua_pushstring(L, sun_params.sunrise.c_str());
+	lua_setfield(L, -2, "sunrise");
+	lua_pushboolean(L, sun_params.sunrise_visible);
+	lua_setfield(L, -2, "sunrise_visible");
+	lua_pushnumber(L, sun_params.rotation);
+	lua_setfield(L, -2, "rotation");
+	lua_pushnumber(L, sun_params.scale);
+	lua_setfield(L, -2, "scale");
+
+	return 1;
+}
+
+// set_moon(self, {visible, texture=, tonemap=, sunrise=, rotation=, scale=})
+int ObjectRef::l_set_moon(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	if (lua_istable(L, 2)) {
+		MoonParams moon_params = player->getMoonParams();
+
+		moon_params.visible = getboolfield_default(L, 2,
+			"visible", moon_params.visible);
+		
+		moon_params.texture = getstringfield_default(L, 2,
+			"texture", moon_params.texture);
+
+		moon_params.tonemap = getstringfield_default(L, 2,
+			"tonemap", moon_params.tonemap);
+		
+		moon_params.rotation = getfloatfield_default(L, 2,
+			"rotation", moon_params.rotation);
+
+		moon_params.scale = getfloatfield_default(L, 2,
+			"scale", moon_params.scale);
+		
+		getServer(L)->setMoon(player, moon_params);
+		return 1;
+	}
+	return 0;
+}
+
+// get_moon(self)
+int ObjectRef::l_get_moon(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	const MoonParams &moon_params = player->getMoonParams();
+
+	lua_newtable(L);
+	lua_pushboolean(L, moon_params.visible);
+	lua_setfield(L, -2, "visible");
+	lua_pushstring(L, moon_params.texture.c_str());
+	lua_setfield(L, -2, "texture");
+	lua_pushstring(L, moon_params.tonemap.c_str());
+	lua_setfield(L, -2, "tonemap");
+	lua_pushnumber(L, moon_params.rotation);
+	lua_setfield(L, -2, "rotation");
+	lua_pushnumber(L, moon_params.scale);
+	lua_setfield(L, -2, "scale");
+
+	return 1;
+}
+
+// set_stars(self, {visible, count=, starcolor=, rotation=, scale=})
+int ObjectRef::l_set_stars(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (!player)
+		return 0;
+	if (lua_istable(L, 2)) {
+		StarParams star_params = player->getStarParams();
+
+		star_params.visible = getboolfield_default(L, 2,
+			"visible", star_params.visible);
+
+		star_params.count = getintfield_default(L, 2,
+			"count", star_params.count);
+
+		star_params.starcolor = video::SColor(255, 235, 235, 255);
```

Comment:
why do default values (such as this one) sometimes appear here?
`RemotePlayer::RemotePlayer()` sets defaults, why does keeping them not work?

Reply:
Defaults are returned from already set values - ie, mod supplied values. If any part was set to `nil`: it'd keep the previously last used value.

See: set_clouds.

Reply:
This should not be necessary, `starcolor` will always be set to default if left alone.
If stuff segfaults if you don't do this, the problem is elsewhere and this code is just bandaid.

Reply:
I believe this was done in the original set_sky, to prevent invalid "colours" being set.

Reply:
`read_color()` has a built in protection which prevents invalid colours like #1234567890

Reply:
@sfan5 is this resolved in your opinion?

Reply:
@Jordach please do not resolve discussions like this, they can only be resolved if the core dev is satisfied.

Reply:
Not resolved. I'd like a concrete case where removing this line breaks stuff.